File: python/rapidsmpf/rapidsmpf/tests/streaming/test_table_chunk.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import TYPE_CHECKING

import cupy
import pytest

import cudf

from rapidsmpf.cuda_stream import is_equal_streams
from rapidsmpf.memory.buffer import MemoryType
from rapidsmpf.memory.content_description import ContentDescription
from rapidsmpf.streaming.core.message import Message
from rapidsmpf.streaming.core.spillable_messages import SpillableMessages
from rapidsmpf.streaming.cudf.table_chunk import TableChunk
from rapidsmpf.testing import assert_eq
from rapidsmpf.utils.cudf import cudf_to_pylibcudf_table

if TYPE_CHECKING:
    import pylibcudf
    from rmm.pylibrmm.stream import Stream

    from rapidsmpf.streaming.core.context import Context


def random_table(nbytes: int) -> pylibcudf.Table:
    assert nbytes % 4 == 0
    return cudf_to_pylibcudf_table(
        cudf.DataFrame(
            {
                "data": cupy.random.random(nbytes // 4, dtype=cupy.float32),
            }
        )
    )


@pytest.mark.parametrize(
    "exclusive_view",
    [True, False],
)
def test_roundtrip(context: Context, stream: Stream, *, exclusive_view: bool) -> None:
    seq = 42
    expect = random_table(1024)
    table_chunk = TableChunk.from_pylibcudf_table(
        expect, stream, exclusive_view=exclusive_view
    )
    assert is_equal_streams(table_chunk.stream, stream)
    assert table_chunk.is_available()
    assert table_chunk.make_available_cost() == 0
    assert table_chunk.is_spillable() == exclusive_view
    assert_eq(expect, table_chunk.table_view())

    # Message roundtrip check.
    msg1 = Message(seq, table_chunk)
    assert msg1.sequence_number == seq
    assert msg1.get_content_description() == ContentDescription(
        content_sizes={
            MemoryType.DEVICE: 1024,
            MemoryType.PINNED_HOST: 0,
            MemoryType.HOST: 0,
        },
        spillable=exclusive_view,
    )

    # Make a copy of msg1 in host memory.
    assert msg1.copy_cost() == 1024
    res, _ = context.br().reserve(MemoryType.HOST, 1024, allow_overbooking=True)
    msg2 = msg1.copy(res)
    assert res.size == 0

    # msg1 is availabe
    table_chunk2 = TableChunk.from_message(msg1)
    assert is_equal_streams(table_chunk2.stream, stream)
    assert table_chunk2.is_available()
    assert table_chunk2.make_available_cost() == 0
    assert_eq(expect, table_chunk2.table_view())

    # Make a copy of msg2 back to device memory.
    assert msg2.copy_cost() == 1024
    res, _ = context.br().reserve(MemoryType.DEVICE, 1024, allow_overbooking=True)
    msg3 = msg2.copy(res)
    assert res.size == 0

    # msg2 is on host and is not availabe
    table_chunk3 = TableChunk.from_message(msg2)
    assert is_equal_streams(table_chunk3.stream, stream)
    assert not table_chunk3.is_available()
    assert table_chunk3.make_available_cost() == 1024
    # but we can make its table available using `make_available()`.
    res, _ = context.br().reserve(MemoryType.DEVICE, 1024, allow_overbooking=True)
    table_chunk4 = table_chunk3.make_available(res)
    assert is_equal_streams(table_chunk4.stream, stream)
    assert table_chunk4.is_available()
    assert table_chunk4.make_available_cost() == 0
    assert_eq(expect, table_chunk4.table_view())

    # msg3 is on device (was created by copying the host msg2). During the copy this
    # is made available trivially.
    table_chunk5 = TableChunk.from_message(msg3)
    assert is_equal_streams(table_chunk5.stream, stream)
    assert table_chunk5.is_available()
    # and it cost no device memory to make available.
    assert table_chunk5.make_available_cost() == 0
    res, _ = context.br().reserve(MemoryType.DEVICE, 0, allow_overbooking=True)
    table_chunk6 = table_chunk5.make_available(res)
    assert table_chunk6.is_available()
    assert table_chunk6.make_available_cost() == 0
    assert_eq(expect, table_chunk6.table_view())


def test_copy_roundtrip(context: Context, stream: Stream) -> None:
    for nrows, ncols in [(1, 1), (1000, 100), (1, 1000)]:
        expect = cudf_to_pylibcudf_table(
            cudf.DataFrame(
                {
                    f"{name}": cupy.random.random(nrows, dtype=cupy.float32)
                    for name in range(ncols)
                }
            )
        )

        tbl1 = TableChunk.from_pylibcudf_table(expect, stream, exclusive_view=True)
        res, _ = context.br().reserve(
            MemoryType.HOST,
            tbl1.data_alloc_size(MemoryType.DEVICE),
            allow_overbooking=True,
        )
        tbl2 = tbl1.copy(res)
        res, _ = context.br().reserve(
            MemoryType.DEVICE, tbl2.make_available_cost(), allow_overbooking=True
        )
        tbl3 = tbl2.make_available(res)
        assert_eq(expect, tbl3.table_view())


def test_spillable_messages(context: Context, stream: Stream) -> None:
    seq = 42
    df1 = random_table(1024)
    df2 = random_table(2048)

    sm = SpillableMessages()
    sm.insert(
        Message(seq, TableChunk.from_pylibcudf_table(df1, stream, exclusive_view=True))
    )
    assert sm.get_content_descriptions() == {
        0: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 1024,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 0,
            },
            spillable=True,
        )
    }
    sm.insert(
        Message(seq, TableChunk.from_pylibcudf_table(df2, stream, exclusive_view=False))
    )
    assert sm.get_content_descriptions() == {
        0: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 1024,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 0,
            },
            spillable=True,
        ),
        1: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 2048,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 0,
            },
            spillable=False,
        ),
    }
    assert sm.spill(mid=0, br=context.br()) == 1024
    assert sm.get_content_descriptions() == {
        0: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 0,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 1024,
            },
            spillable=True,
        ),
        1: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 2048,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 0,
            },
            spillable=False,
        ),
    }
    assert sm.spill(mid=1, br=context.br()) == 0
    assert sm.get_content_descriptions() == {
        0: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 0,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 1024,
            },
            spillable=True,
        ),
        1: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 2048,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 0,
            },
            spillable=False,
        ),
    }

    # Extract, make available, and check table chunk 1.
    df1_got = TableChunk.from_message(sm.extract(mid=0))
    res, _ = context.br().reserve(
        MemoryType.DEVICE, df1_got.make_available_cost(), allow_overbooking=True
    )
    df1_got = df1_got.make_available(res)
    assert_eq(df1, df1_got.table_view())

    with pytest.raises(IndexError, match="Invalid key"):
        sm.extract(mid=0)

    df2_got = TableChunk.from_message(sm.extract(mid=1))
    df2_got = df2_got.make_available_and_spill(context.br(), allow_overbooking=True)
    assert_eq(df2, df2_got.table_view())
    assert sm.get_content_descriptions() == {}


def test_spillable_messages_by_context(context: Context, stream: Stream) -> None:
    seq = 42
    expect = random_table(1024)

    mid = context.spillable_messages().insert(
        Message(
            seq, TableChunk.from_pylibcudf_table(expect, stream, exclusive_view=True)
        )
    )
    assert context.spillable_messages().get_content_descriptions() == {
        0: ContentDescription(
            content_sizes={
                MemoryType.DEVICE: 1024,
                MemoryType.PINNED_HOST: 0,
                MemoryType.HOST: 0,
            },
            spillable=True,
        )
    }
    got = TableChunk.from_message(context.spillable_messages().extract(mid=mid))
    assert_eq(expect, got.table_view())

File: cpp/tests/streaming/test_table_chunk.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */


#include <cstdint>
#include <memory>

#include <gtest/gtest.h>

#include <cudf/contiguous_split.hpp>
#include <cudf/table/table.hpp>
#include <cudf/table/table_view.hpp>
#include <cudf_test/column_wrapper.hpp>
#include <cudf_test/table_utilities.hpp>
#include <rmm/cuda_stream_view.hpp>
#include <rmm/mr/per_device_resource.hpp>

#include <rapidsmpf/owning_wrapper.hpp>
#include <rapidsmpf/streaming/core/channel.hpp>
#include <rapidsmpf/streaming/cudf/table_chunk.hpp>

#include "../utils.hpp"
#include "base_streaming_fixture.hpp"


using namespace rapidsmpf;
using namespace rapidsmpf::streaming;

class StreamingTableChunk : public BaseStreamingFixture,
                            public ::testing::WithParamInterface<rapidsmpf::MemoryType> {
  protected:
    void SetUp() override {
        rapidsmpf::config::Options options(
            rapidsmpf::config::get_environment_variables()
        );

        std::unordered_map<MemoryType, rapidsmpf::BufferResource::MemoryAvailable>
            memory_available{};
        auto stream_pool = std::make_shared<rmm::cuda_stream_pool>(
            16, rmm::cuda_stream::flags::non_blocking
        );
        stream = cudf::get_default_stream();
        br = std::make_shared<rapidsmpf::BufferResource>(
            mr_cuda,  // device_mr
            rapidsmpf::PinnedMemoryResource::make_if_available(),  // pinned_mr
            memory_available,  // memory_available
            std::chrono::milliseconds{1},  // periodic_spill_check
            stream_pool,  // stream_pool
            Statistics::disabled()  // statistics
        );
        ctx = std::make_shared<rapidsmpf::streaming::Context>(
            options, GlobalEnvironment->comm_, br
        );
    }

    rmm::cuda_stream_view stream;
    rmm::mr::cuda_memory_resource mr_cuda;
    std::shared_ptr<rapidsmpf::BufferResource> br;
    std::shared_ptr<rapidsmpf::streaming::Context> ctx;
};

TEST_F(StreamingTableChunk, FromTable) {
    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;

    cudf::table expect = random_table_with_index(seed, num_rows, 0, 10);

    TableChunk chunk{std::make_unique<cudf::table>(expect), stream};
    EXPECT_EQ(chunk.stream().value(), stream.value());
    EXPECT_TRUE(chunk.is_available());
    EXPECT_TRUE(chunk.is_spillable());
    EXPECT_EQ(chunk.make_available_cost(), 0);
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(chunk.table_view(), expect);
}

TEST_F(StreamingTableChunk, TableChunkOwner) {
    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;
    constexpr std::uint64_t seq = 42;

    cudf::table expect = random_table_with_index(seed, num_rows, 0, 10);
    // Static because the deleter function is a void(*)(void*) which precludes the use of
    // a lambda with captures.
    static std::size_t num_deletions{0};
    auto deleter = [](void* p) {
        num_deletions++;
        delete static_cast<int*>(p);
    };
    auto make_chunk = [&](TableChunk::ExclusiveView exclusive_view) {
        return TableChunk{
            expect, stream, OwningWrapper(new int, deleter), exclusive_view
        };
    };
    auto check_chunk = [&](TableChunk const& chunk, bool is_spillable) {
        EXPECT_EQ(chunk.stream().value(), stream.value());
        EXPECT_TRUE(chunk.is_available());
        EXPECT_EQ(chunk.is_spillable(), is_spillable);
        EXPECT_EQ(chunk.make_available_cost(), 0);
        CUDF_TEST_EXPECT_TABLES_EQUIVALENT(chunk.table_view(), expect);
    };
    {
        auto chunk = make_chunk(TableChunk::ExclusiveView::NO);
        check_chunk(chunk, false);
        EXPECT_EQ(num_deletions, 0);
    }
    EXPECT_EQ(num_deletions, 1);
    {
        auto msg = to_message(
            seq, std::make_unique<TableChunk>(make_chunk(TableChunk::ExclusiveView::NO))
        );
        EXPECT_EQ(num_deletions, 1);
    }
    EXPECT_EQ(num_deletions, 2);
    {
        auto msg = to_message(
            seq, std::make_unique<TableChunk>(make_chunk(TableChunk::ExclusiveView::YES))
        );
        auto chunk = msg.release<TableChunk>();
        check_chunk(chunk, true);
        EXPECT_EQ(num_deletions, 2);
    }
    EXPECT_EQ(num_deletions, 3);
    {
        auto chunk = make_chunk(TableChunk::ExclusiveView::YES);
        check_chunk(chunk, true);
        auto res = br->reserve_or_fail(
            chunk.data_alloc_size(MemoryType::DEVICE), MemoryType::DEVICE
        );
        // This is like spilling since the original `chunk` is ExclusiveView::YES and
        // overwritten.
        chunk = chunk.copy(res);
        EXPECT_EQ(num_deletions, 4);
    }
}

TEST_F(StreamingTableChunk, FromPackedColumns) {
    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;

    cudf::table expect = random_table_with_index(seed, num_rows, 0, 10);
    auto packed = cudf::pack(expect, stream);

    TableChunk chunk{std::make_unique<cudf::packed_columns>(std::move(packed)), stream};

    EXPECT_EQ(chunk.stream().value(), stream.value());
    EXPECT_TRUE(chunk.is_available());
    EXPECT_TRUE(chunk.is_spillable());
    EXPECT_EQ(chunk.make_available_cost(), 0);
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(chunk.table_view(), expect);
}

TEST_F(StreamingTableChunk, FromPackedDataOnDevice) {
    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;

    cudf::table expect = random_table_with_index(seed, num_rows, 0, 10);
    auto packed_columns = cudf::pack(expect, stream);

    auto packed_data = std::make_unique<PackedData>(
        std::move(packed_columns.metadata),
        br->move(std::move(packed_columns.gpu_data), stream)
    );
    TableChunk chunk{std::move(packed_data)};

    EXPECT_EQ(chunk.stream().value(), stream.value());
    EXPECT_FALSE(chunk.is_available());
    EXPECT_TRUE(chunk.is_spillable());
    EXPECT_THROW((void)chunk.table_view(), std::invalid_argument);

    // Eventhough the table isn't available, it is still all in device memory
    // so the memory cost of making it available is zero.
    EXPECT_EQ(chunk.make_available_cost(), 0);
}

INSTANTIATE_TEST_SUITE_P(
    StreamingTableChunkWithSpillTargets,
    StreamingTableChunk,
    ::testing::ValuesIn(rapidsmpf::SPILL_TARGET_MEMORY_TYPES),
    [](testing::TestParamInfo<rapidsmpf::MemoryType> const& info) {
        return std::string{rapidsmpf::to_string(info.param)};
    }
);

TEST_P(StreamingTableChunk, FromPackedDataOn) {
    auto const spill_mem_type = GetParam();
    if (spill_mem_type == MemoryType::PINNED_HOST
        && !is_pinned_memory_resources_supported())
    {
        GTEST_SKIP() << "MemoryType::PINNED_HOST isn't supported on the system.";
    }

    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;

    cudf::table expect = random_table_with_index(seed, num_rows, 0, 10);
    auto packed_columns = cudf::pack(expect, stream);
    std::size_t const size = packed_columns.gpu_data->size();

    // Move the gpu_data to a Buffer (still device memory).
    auto gpu_data_on_device = br->move(std::move(packed_columns.gpu_data), stream);

    // Copy the GPU data to the current spill target memory type.
    auto [res, _] = br->reserve(spill_mem_type, size, true);
    auto gpu_data_in_spill_memory = br->move(std::move(gpu_data_on_device), res);

    auto packed_data = std::make_unique<PackedData>(
        std::move(packed_columns.metadata), std::move(gpu_data_in_spill_memory)
    );
    TableChunk chunk{std::move(packed_data)};

    EXPECT_EQ(chunk.stream().value(), stream.value());
    EXPECT_FALSE(chunk.is_available());
    EXPECT_TRUE(chunk.is_spillable());
    EXPECT_THROW((void)chunk.table_view(), std::invalid_argument);
    EXPECT_EQ(chunk.make_available_cost(), size);
}

TEST_F(StreamingTableChunk, DeviceToDeviceCopy) {
    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;

    auto expect = random_table_with_index(seed, num_rows, 0, 10);

    rapidsmpf::streaming::TableChunk chunk{std::make_unique<cudf::table>(expect), stream};
    EXPECT_TRUE(chunk.is_available());

    auto res = br->reserve_or_fail(
        chunk.data_alloc_size(MemoryType::DEVICE), MemoryType::DEVICE
    );
    auto chunk2 = chunk.copy(res);

    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(chunk2.table_view(), expect);
}

TEST_P(StreamingTableChunk, DeviceToHostRoundTripCopy) {
    auto const spill_mem_type = GetParam();
    if (spill_mem_type == MemoryType::PINNED_HOST
        && !is_pinned_memory_resources_supported())
    {
        GTEST_SKIP() << "MemoryType::PINNED_HOST isn't supported on the system.";
    }

    constexpr unsigned int num_rows = 64;
    constexpr std::int64_t seed = 2025;

    auto expect = random_table_with_index(seed, num_rows, 0, 5);

    TableChunk dev_chunk{std::make_unique<cudf::table>(expect), stream};
    EXPECT_TRUE(dev_chunk.is_available());
    EXPECT_TRUE(dev_chunk.is_spillable());
    EXPECT_EQ(dev_chunk.stream().value(), stream.value());
    EXPECT_EQ(dev_chunk.make_available_cost(), 0);
    {
        auto cd = get_content_description(dev_chunk);
        EXPECT_EQ(cd.spillable(), dev_chunk.is_spillable());
        for (auto mem_type : MEMORY_TYPES) {
            EXPECT_EQ(cd.content_size(mem_type), dev_chunk.data_alloc_size(mem_type));
        }
    }

    // Copy to host memory -> new chunk should be unavailable.
    auto host_res = br->reserve_or_fail(
        dev_chunk.data_alloc_size(MemoryType::DEVICE), spill_mem_type
    );
    auto host_copy = dev_chunk.copy(host_res);
    EXPECT_FALSE(host_copy.is_available());
    EXPECT_TRUE(host_copy.is_spillable());
    EXPECT_EQ(host_copy.stream().value(), stream.value());
    EXPECT_GT(host_copy.make_available_cost(), 0);
    {
        auto cd = get_content_description(host_copy);
        EXPECT_EQ(cd.spillable(), host_copy.is_spillable());
        for (auto mem_type : MEMORY_TYPES) {
            EXPECT_EQ(cd.content_size(mem_type), host_copy.data_alloc_size(mem_type));
        }
    }

    // Host to host copy.
    auto host_res2 =
        br->reserve_or_fail(host_copy.data_alloc_size(spill_mem_type), spill_mem_type);
    auto host_copy2 = host_copy.copy(host_res2);
    EXPECT_FALSE(host_copy2.is_available());
    EXPECT_TRUE(host_copy2.is_spillable());
    EXPECT_EQ(host_copy2.stream().value(), stream.value());
    EXPECT_EQ(host_copy2.make_available_cost(), host_copy.make_available_cost());
    {
        auto cd = get_content_description(host_copy2);
        EXPECT_EQ(cd.spillable(), host_copy2.is_spillable());
        for (auto mem_type : MEMORY_TYPES) {
            EXPECT_EQ(cd.content_size(mem_type), host_copy2.data_alloc_size(mem_type));
        }
    }

    // Bring the new host copy back to device and verify equality.
    auto dev_res = br->reserve_or_fail(
        host_copy2.data_alloc_size(spill_mem_type), MemoryType::DEVICE
    );
    auto dev_back = host_copy2.make_available(dev_res);
    EXPECT_TRUE(dev_back.is_available());
    EXPECT_TRUE(dev_back.is_spillable());
    EXPECT_EQ(dev_back.stream().value(), stream.value());
    EXPECT_EQ(dev_back.make_available_cost(), 0);
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(dev_back.table_view(), expect);
    {
        auto cd = get_content_description(dev_back);
        EXPECT_EQ(cd.spillable(), dev_back.is_spillable());
        for (auto mem_type : MEMORY_TYPES) {
            EXPECT_EQ(cd.content_size(mem_type), dev_back.data_alloc_size(mem_type));
        }
    }

    // Sanity check: a second device copy should also remain equivalent.
    auto dev_res2 = br->reserve_or_fail(
        dev_back.data_alloc_size(MemoryType::DEVICE), MemoryType::DEVICE
    );
    auto dev_copy2 = dev_back.copy(dev_res2);
    EXPECT_TRUE(dev_copy2.is_available());
    EXPECT_EQ(dev_copy2.make_available_cost(), 0);
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(dev_copy2.table_view(), expect);
    {
        auto cd = get_content_description(dev_copy2);
        EXPECT_EQ(cd.spillable(), dev_copy2.is_spillable());
        for (auto mem_type : MEMORY_TYPES) {
            EXPECT_EQ(cd.content_size(mem_type), dev_copy2.data_alloc_size(mem_type));
        }
    }
}

TEST_F(StreamingTableChunk, ToMessageRoundTrip) {
    constexpr unsigned int num_rows = 64;
    constexpr std::int64_t seed = 2025;
    constexpr std::uint64_t seq = 7;

    auto expect = random_table_with_index(seed, num_rows, 0, 5);
    auto chunk =
        std::make_unique<TableChunk>(std::make_unique<cudf::table>(expect), stream);

    Message m = to_message(seq, std::move(chunk));
    EXPECT_FALSE(m.empty());
    EXPECT_TRUE(m.holds<TableChunk>());
    EXPECT_TRUE(m.content_description().spillable());
    EXPECT_EQ(m.content_description().content_size(MemoryType::HOST), 0);
    EXPECT_EQ(m.content_description().content_size(MemoryType::DEVICE), 1024);
    EXPECT_EQ(m.sequence_number(), seq);

    // Deep-copy: device to host.
    auto reservation = br->reserve_or_fail(m.copy_cost(), MemoryType::HOST);
    Message m2 = m.copy(reservation);
    EXPECT_EQ(reservation.size(), 0);
    EXPECT_FALSE(m2.empty());
    EXPECT_TRUE(m2.holds<TableChunk>());
    EXPECT_TRUE(m2.content_description().spillable());
    EXPECT_EQ(m2.content_description().content_size(MemoryType::HOST), 1024);
    EXPECT_EQ(m2.content_description().content_size(MemoryType::DEVICE), 0);
    EXPECT_EQ(m2.sequence_number(), seq);

    // Deep-copy: host to host.
    reservation = br->reserve_or_fail(m2.copy_cost(), MemoryType::HOST);
    Message m3 = m.copy(reservation);
    EXPECT_EQ(reservation.size(), 0);
    EXPECT_FALSE(m3.empty());
    EXPECT_TRUE(m3.holds<TableChunk>());
    EXPECT_TRUE(m3.content_description().spillable());
    EXPECT_EQ(m3.content_description().content_size(MemoryType::HOST), 1024);
    EXPECT_EQ(m3.content_description().content_size(MemoryType::DEVICE), 0);
    EXPECT_EQ(m3.sequence_number(), seq);

    // Copy the chunk back to device and verify.
    {
        auto chunk = m3.release<TableChunk>();
        auto res = br->reserve_or_fail(chunk.make_available_cost(), MemoryType::DEVICE);
        chunk = chunk.make_available(res);
        CUDF_TEST_EXPECT_TABLES_EQUIVALENT(chunk.table_view(), expect);
    }

    // Deep-copy: host to device.
    reservation = br->reserve_or_fail(m2.copy_cost(), MemoryType::DEVICE);
    Message m4 = m.copy(reservation);
    EXPECT_EQ(reservation.size(), 0);
    EXPECT_FALSE(m4.empty());
    EXPECT_TRUE(m4.holds<TableChunk>());
    EXPECT_TRUE(m4.content_description().spillable());
    EXPECT_EQ(m4.content_description().content_size(MemoryType::HOST), 0);
    EXPECT_EQ(m4.content_description().content_size(MemoryType::DEVICE), 1024);
    EXPECT_EQ(m4.sequence_number(), seq);
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(m4.get<TableChunk>().table_view(), expect);

    // Deep-copy: device to device.
    reservation = br->reserve_or_fail(m4.copy_cost(), MemoryType::DEVICE);
    Message m5 = m.copy(reservation);
    EXPECT_EQ(reservation.size(), 0);
    EXPECT_FALSE(m5.empty());
    EXPECT_TRUE(m5.holds<TableChunk>());
    EXPECT_TRUE(m5.content_description().spillable());
    EXPECT_EQ(m5.content_description().content_size(MemoryType::HOST), 0);
    EXPECT_EQ(m5.content_description().content_size(MemoryType::DEVICE), 1024);
    EXPECT_EQ(m5.sequence_number(), seq);
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(m5.get<TableChunk>().table_view(), expect);
}

TEST_F(StreamingTableChunk, ToMessageNotSpillable) {
    constexpr unsigned int num_rows = 100;
    constexpr std::int64_t seed = 1337;
    constexpr std::uint64_t seq = 42;

    cudf::table expect = random_table_with_index(seed, num_rows, 0, 10);

    auto deleter = [](void* p) { delete static_cast<int*>(p); };
    auto chunk = std::make_unique<TableChunk>(
        expect, stream, OwningWrapper(new int, deleter), TableChunk::ExclusiveView::NO
    );

    Message m = to_message(seq, std::move(chunk));
    EXPECT_FALSE(m.empty());
    EXPECT_TRUE(m.holds<TableChunk>());
    EXPECT_FALSE(m.content_description().spillable());
    EXPECT_EQ(m.content_description().content_size(MemoryType::HOST), 0);
    EXPECT_EQ(
        m.content_description().content_size(MemoryType::DEVICE), expect.alloc_size()
    );
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(m.get<TableChunk>().table_view(), expect);
}

TEST_F(StreamingTableChunk, ToMessageUnalignedSize) {
    constexpr unsigned int num_rows = 5;
    constexpr std::int64_t seed = 2025;
    constexpr std::uint64_t seq = 7;

    auto expect = random_table_with_index(seed, num_rows, 0, 5);
    auto chunk =
        std::make_unique<TableChunk>(std::make_unique<cudf::table>(expect), stream);

    Message m = to_message(seq, std::move(chunk));
    EXPECT_EQ(m.sequence_number(), seq);
    EXPECT_FALSE(m.empty());
    EXPECT_TRUE(m.holds<TableChunk>());
    EXPECT_TRUE(m.content_description().spillable());
    EXPECT_EQ(m.content_description().content_size(MemoryType::HOST), 0);
    EXPECT_EQ(m.content_description().content_size(MemoryType::DEVICE), 80);
    EXPECT_EQ(m.copy_cost(), 80);

    // Deep copy: device → host.
    // Note: `m.copy_cost() == 80`, but cudf performs 128-byte aligned allocations.
    // This means `m.copy_cost()` is not always sufficient; however, TableChunk.copy()
    // accounts for this alignment internally.
    auto reservation = br->reserve_or_fail(m.copy_cost(), MemoryType::HOST);
    Message m2 = m.copy(reservation);
    EXPECT_EQ(reservation.size(), 0);
    EXPECT_FALSE(m2.empty());
    EXPECT_TRUE(m2.holds<TableChunk>());
    EXPECT_TRUE(m2.content_description().spillable());
    EXPECT_EQ(m2.copy_cost(), 128);
    EXPECT_EQ(m2.content_description().content_size(MemoryType::HOST), 128);
    EXPECT_EQ(m2.content_description().content_size(MemoryType::DEVICE), 0);
    EXPECT_EQ(m2.sequence_number(), seq);
}
